util/io_posix.cc

```diff
@@ -730,11 +730,13 @@ Status PosixWritableFile::PositionedAppend(const Slice& data, uint64_t offset) {
 
 Status PosixWritableFile::Truncate(uint64_t size) {
   Status s;
-  int r = ftruncate(fd_, size);
-  if (r < 0) {
-    s = IOError(filename_, errno);
-  } else {
-    filesize_ = size;
+  if (use_direct_io()) {
```

Comment:
I think this logic should happen in WritableFileWriter. It's more straight-forward if File::Truncate() always truncates.